### PR TITLE
[mdns] keep avahi service resolver until instance is removed

### PR DESCRIPTION
**Pull request by sunytt**
There is a difference between avahi and dns-sd browse operation. An update in a discovered instance doesn't trigger any avahi browse events, while for dns-sd rmv/add events are triggered.

In the current avahi mdns design, after service browse we resolve the discovered instance once, when instance is resolved the resolver is removed.

Trel uses dnssd browse to discover peers, when a discovered peer changes ipv6 address, trel would not know the change and still try to send packets to the old address.

This PR changes the avahi mdns to keep service resolver until the instance is removed, so any updates to the instance will be noticed.

A followup to this PR is to add test in openthread.

### CODE REVIEW

File: src/mdns/mdns_avahi.cpp [hunk 1]
```diff
@@ -1176,6 +1175,10 @@ void PublisherAvahi::ServiceSubscription::HandleResolveResult(AvahiServiceResolv
 exit:
     if (resolved)
     {
+        if (instanceInfo.mRemoved)
```
Comment (superwhd) [inline src/mdns/mdns_avahi.cpp:1178]: When will this happen?

File: src/mdns/mdns_avahi.cpp [hunk 2]
```diff
@@ -1050,6 +1054,7 @@ void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser
         break;
     case AVAHI_BROWSER_REMOVE:
         mPublisherAvahi->OnServiceRemoved(static_cast<uint32_t>(aInterfaceIndex), aType, aName);
+        RemoveServiceResolver(aName);
```
Comment (superwhd) [inline src/mdns/mdns_avahi.cpp:1057]: Is it able to handle this scenario?

```
1. service A is published.
2. service A is unpublished.
3. service A is published again.
4. service A is updated.
```

File: src/mdns/mdns_avahi.cpp [hunk 3]
```diff
@@ -1185,19 +1189,32 @@ void PublisherAvahi::ServiceSubscription::HandleResolveResult(AvahiServiceResolv
...
+    otbrLogDebug("Added service resolver for instance %s", aInstanceName.c_str());
 }
 
-void PublisherAvahi::ServiceSubscription::RemoveServiceResolver(AvahiServiceResolver *aServiceResolver)
+void PublisherAvahi::ServiceSubscription::RemoveServiceResolver(std::string aInstanceName)
 {
-    assert(aServiceResolver != nullptr);
-    assert(mServiceResolvers.find(aServiceResolver) != mServiceResolvers.end());
+    int numResolvers = 0;
+    VerifyOrExit(mServiceResolvers.find(aInstanceName) != mServiceResolvers.end());
```
Comment (superwhd) [inline src/mdns/mdns_avahi.cpp:1204]:
```suggestion
    int numResolvers = 0;

    VerifyOrExit(mServiceResolvers.find(aInstanceName) != mServiceResolvers.end());
```

File: src/mdns/mdns_avahi.hpp [hunk 4]
```diff
@@ -184,8 +184,8 @@ class PublisherAvahi : public Publisher
                      AvahiProtocol      aProtocol,
                      const std::string &aInstanceName,
                      const std::string &aType);
-        void AddServiceResolver(AvahiServiceResolver *aServiceResolver);
-        void RemoveServiceResolver(AvahiServiceResolver *aServiceResolver);
+        void AddServiceResolver(std::string, AvahiServiceResolver *aServiceResolver);
+        void RemoveServiceResolver(std::string aInstanceName);
```
Comment (superwhd) [inline src/mdns/mdns_avahi.hpp:188]:
```suggestion
        void AddServiceResolver(const std::string &aInstanceName, AvahiServiceResolver *aServiceResolver);
        void RemoveServiceResolver(const std::string &aInstanceName);
```

File: src/mdns/mdns_avahi.cpp (hunk 2 above)
Comment (sunytt) [reply src/mdns/mdns_avahi.cpp:1057]: The scenario works, tested with 2 devices, device 1 with trel disabled, device 2 with trel enabled:
1. enable trel on device 1, service A is published and resolved by device 2
2. disable trel on device 1, service A is removed and resolved by device 2
3. enable trel on device 1, service A is added and resolved by device 2
4. reboot device 1, service A is updated and resolved by device 2

logs on device B:
```
1)
03-27 02:21:19.969  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.977  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.985  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.994  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.995  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1
03-27 02:21:20.000  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1

2)
03-27 02:22:16.881  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.897  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 2 service resolver for instance 12e24f966ff2295c
03-27 02:22:16.897  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.907  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 2 service resolver for instance 12e24f966ff2295c
03-27 02:22:16.907  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.907  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 0 service resolver for instance 12e24f966ff2295c
03-27 02:22:16.908  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.908  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 0 service resolver for instance 12e24f966ff2295c

3)
03-27 02:22:44.642  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.651  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.658  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.666  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.666  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1
03-27 02:22:44.671  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1

4)
03-27 02:24:59.952  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host CA1750BED820.local. addresses 1
03-27 02:24:59.959  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host CA1750BED820.local. addresses 1
```

Comment (sunytt) [reply src/mdns/mdns_avahi.cpp:1057]: Offline discussed with @superwhd , `avahi_service_browser_new` uses `AVAHI_PROTO_UNSPEC ` while `avahi_service_resolver_new` uses `AVAHI_PROTO_INET6 `, there might be cases when browse callback was called for v4 event while resolver is created with v6 only.

Updated `avahi_service_resolver_new` to use `AVAHI_PROTO_UNSPEC ` to align with browser, and the logs for above scenario becomes:

```
1)
03-27 10:01:45.311  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:01:45.319  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:01:45.326  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 2E05397250BE-2.local. addresses 1

2)
03-27 10:02:13.548  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 10:02:13.556  1986  1986 D otbr-agent: [DEBG]-MDNS----: Removed 2 service resolver for instance 12e24f966ff2295c
03-27 10:02:13.556  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 10:02:13.557  1986  1986 D otbr-agent: [DEBG]-MDNS----: Removed 0 service resolver for instance 12e24f966ff2295c

3)
03-27 10:02:19.909  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:02:19.917  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:02:19.922  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 2E05397250BE-2.local. addresses 1

4)
03-27 10:03:39.284  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 9ED50C0C67C0.local. addresses 1
```

Comment (wgtdkp) [reply src/mdns/mdns_avahi.cpp:1057]: > Offline discussed with @superwhd , avahi_service_browser_new uses AVAHI_PROTO_UNSPEC  while avahi_service_resolver_new uses AVAHI_PROTO_INET6 , there might be cases when browse callback was called for v4 event while resolver is created with v6 only.
>
>Updated avahi_service_resolver_new to use AVAHI_PROTO_UNSPEC  to align with browser, and the logs for above scenario becomes:

+1, it's recommended to browse/resolve with both IPv4 and IPv6 per the Discovery Proxy spec:
```
At present, it is RECOMMENDED that a Discovery Proxy be configured with a single domain name for both
the IPv4 and IPv6 ".local." zones on the local link, and when a unicast query is received, it should issue
Multicast DNS queries using both IPv4 and IPv6 on the local link and then combine the results.
```

**Review by superwhd — APPROVED** [review]
LGTM 👍

**Review by wgtdkp — APPROVED** [review]
LGTM :+1: 
Thanks!

**Review by jwhui — APPROVED** [review]
Thanks! 👍🏼

**Pull request merged**